File: src/main/java/weld/WeldStruct.java
package weld;

import weld.types.PrimitiveType;
import weld.types.StructType;
import weld.types.Type;
import weld.types.VecType;

/**
 * Wrapper for a weld struct value.
 */
public final class WeldStruct extends WeldObject implements AutoCloseable {
  private final StructType type;

  /**
   * Flag to indicate that we should free the object when close() is called
   */
  private boolean freeOnClose;

  /**
   * Create a weld struct that points to a memory address.
   */
  public WeldStruct(long pointer, StructType type) {
    this(pointer, type, false);
  }

  /**
   * Create a weld struct that points to a memory address and that may optionally be freed when
   * close() is called.
   */
  private WeldStruct(long pointer, StructType type, boolean freeOnClose) {
    super(pointer);
    this.type = type;
    this.freeOnClose = freeOnClose;
  }

  @Override
  public long size() {
    return type.size();
  }

  @Override
  public StructType type() {
    return type;
  }

  @Override
  long indexToAddress(int index) {
    return address() + type.fieldOffset(index);
  }

  @Override
  public long numElements() {
    return type.numFields();
  }

  @Override
  public Type getElementType(int index) {
    return type.fieldType(index);
  }

  /**
   * Close the weld struct, it should not be used after this call as it will try free the
   * memory location it points to.
   */
  @Override
  public synchronized void close() {
    if (freeOnClose) {
      Platform.freeMemory(address());
      freeOnClose = false;
    }
  }

  /**
   * Turn the weld struct into an array of values.
   */
  public Object[] toArray() {
    final Object[] result = new Object[(int) numElements()];
    for (int i = 0; i < numElements(); i++) {
      final Type fieldType = type.fieldType(i);
      if (fieldType == PrimitiveType.bool) {
        result[i] = getBoolean(i);
      } else if (fieldType == PrimitiveType.i8) {
        result[i] = getByte(i);
      } else if (fieldType == PrimitiveType.i32) {
        result[i] = getInt(i);
      } else if (fieldType == PrimitiveType.i64) {
        result[i] = getLong(i);
      } else if (fieldType == PrimitiveType.f32) {
        result[i] = getFloat(i);
      } else if (fieldType == PrimitiveType.f64) {
        result[i] = getDouble(i);
      } else if (fieldType == PrimitiveType.Pointer) {
        result[i] = getPointer(i);
      } else if (fieldType instanceof VecType) {
        result[i] = getVec(i);
      } else {
        throw new IllegalArgumentException("Unsupported struct field type[" + i + "]: " + fieldType);
      }
    }
    return result;
  }

  /**
   * Convert the struct into a WeldValue.
   */
  public WeldValue toValue() {
    return new WeldValue(address(), size());
  }

  /**
   * Create a struct from an argument array. Note that structs are not support.
   */
  public static WeldStruct struct(final Object... values) {
    // Determine the struct type.
    // Note that we need to align the offset to the size vecOf the value vecOf we are about to write.
    final Type[] fieldTypes = new Type[values.length];
    long vectorSize = 0;
    for (int i = 0; i < values.length; i++) {
      final Object value = values[i];
      if (value instanceof Boolean) {
        fieldTypes[i] = PrimitiveType.bool;
      } else if (value instanceof Byte) {
        fieldTypes[i] = PrimitiveType.i8;
      } else if (value instanceof Integer) {
        fieldTypes[i] = PrimitiveType.i32;
      } else if (value instanceof Float) {
        fieldTypes[i] = PrimitiveType.f32;
      } else if (value instanceof Long) {
        fieldTypes[i] = PrimitiveType.i64;
      } else if (value instanceof Double) {
        fieldTypes[i] = PrimitiveType.f64;
      } else if (value instanceof WeldVec) {
        fieldTypes[i] = ((WeldVec) value).type();
      } else if (value instanceof WeldVec.Builder) {
        final WeldVec.Builder builder = (WeldVec.Builder) value;
        fieldTypes[i] = VecType.vecOf(builder.type);
        vectorSize += ceil8(builder.size());
      } else {
        throw new IllegalArgumentException("Unsupported struct value[" + i + "]: " + value);
      }
    }
    final StructType type = StructType.structOf(fieldTypes);

    // Create the struct
    final long address = Platform.allocateMemory(type.size() + vectorSize);
    long dataOffset = type.size();
    for (int i = 0; i < values.length; i++) {
      final Object value = values[i];
      final long fieldAddress = address + type.fieldOffset(i);
      if (value instanceof Boolean) {
        final boolean condition = (Boolean) value;
        Platform.putByte(fieldAddress, condition ? (byte) 1 : (byte) 0);
      } else if (value instanceof Byte) {
        Platform.putByte(fieldAddress, (Byte) value);
      } else if (value instanceof Integer) {
        Platform.putInt(fieldAddress, (Integer) value);
      } else if (value instanceof Float) {
        Platform.putFloat(fieldAddress, (Float) value);
      } else if (value instanceof Long) {
        Platform.putLong(fieldAddress, (Long) value);
      } else if (value instanceof Double) {
        Platform.putDouble(fieldAddress, (Double) value);
      } else if (value instanceof WeldVec) {
        final WeldVec vec = (WeldVec) value;
        Platform.putLong(fieldAddress, vec.address());
        Platform.putLong(fieldAddress + 8, vec.numElements());
      } else if (value instanceof WeldVec.Builder) {
        final WeldVec.Builder builder = (WeldVec.Builder) value;
        Platform.putLong(fieldAddress, address + dataOffset);
        Platform.putLong(fieldAddress + 8, builder.numElements);
        builder.putValues(address + dataOffset);
        dataOffset += ceil8(builder.size());
      }
    }
    final WeldStruct result = new WeldStruct(address, type, true);
    Platform.registerForCleanUp(result);
    return result;
  }

  /**
   * Round an int up to its nearest multiple vecOf 8.
   */
  private static long ceil8(long value) {
    return (value + 0x7L) & ~0x7L;
  }
}

File: src/main/scala/weld/WeldStruct.scala
package weld

import scala.annotation.varargs

/**
 * Wrapper for a weld struct value.
 */
class WeldStruct(
    address: Long,
    val objectType: StructType,
    private var freeOnClose: Boolean = false)
  extends WeldObject(address)
  with AutoCloseable {

  override def size: Long = objectType.size
  override def numElements: Long = objectType.fields.size
  override def getElementType(index: Int): WeldType = objectType.fields(index).fieldType
  override private[weld] def indexToAddress(index: Int) = address + objectType.fields(index).offset

  /**
   * Turn the weld struct into an array of values.
   */
  def toArray: Array[Any] = {
    objectType.fields.zipWithIndex.toArray.map {
      case (FieldInfo(fieldType, _), i) =>
        fieldType match {
          case `bool` => getBoolean(i)
          case `i8` => getByte(i)
          case `i16` => getShort(i)
          case `u16` => getChar(i)
          case `i32` => getInt(i)
          case `i64` => getLong(i)
          case `f32` => getFloat(i)
          case `f64` => getDouble(i)
          case Pointer => getPointer(i)
          case _: VecType => getVec(i)
          // Dictionaries are pointers - represent them as such.
          case _: DictMerger => getPointer(i)
          case _: GroupMerger => getPointer(i)
          case _ =>
            throw new IllegalArgumentException(s"Unsupported struct field type[$i]: $fieldType")
        }
    }
  }

  /**
   * Convert the struct into a WeldValue.
   */
  def toValue = WeldValue(address, size)

  private var autoClean: Boolean = false

  def markAutoCleanable(): Unit = {
    if (freeOnClose && !autoClean) {
      autoClean = true
      Platform.registerForCleanUp(this, new WeldStruct.Cleaner(address))
    }
  }

  /**
   * Close the weld struct, it should not be used after this call as it will try free the
   * memory location it points to.
   */
  override def close(): Unit = {
    if (freeOnClose && !autoClean) {
      Platform.freeMemory(address)
      freeOnClose = false
    }
  }
}

object WeldStruct {
  def zero(structType: StructType): WeldStruct = {
    val address = Platform.allocateMemory(structType.size)
    structType.fields.foreach {
      case FieldInfo(`bool`, offset) =>
        Platform.putByte(address + offset, 0.toByte)
      case FieldInfo(`i8`, offset) =>
        Platform.putByte(address + offset, 0.toByte)
      case FieldInfo(`i16`, offset) =>
        Platform.putShort(address + offset, 0.toShort)
      case FieldInfo(`u16`, offset) =>
        Platform.putChar(address + offset, 0.toChar)
      case FieldInfo(`i32`, offset) =>
        Platform.putInt(address + offset, 0)
      case FieldInfo(`i64`, offset) =>
        Platform.putLong(address + offset, 0L)
      case FieldInfo(`f32`, offset) =>
        Platform.putFloat(address + offset, 0f)
      case FieldInfo(`f64`, offset) =>
        Platform.putDouble(address + offset, 0d)
      case FieldInfo(_: PointerType, offset) =>
        Platform.putLong(address + offset, 0L)
      case FieldInfo(_: VecType, offset) =>
        Platform.putLong(address + offset, 0L)
        Platform.putLong(address + offset + 8, 0L)
      case FieldInfo(dt, _) =>
        Platform.freeMemory(address)
        throw new IllegalArgumentException(s"Cannot create an initial value for data type: $dt")
    }
    new WeldStruct(address, structType, freeOnClose = true)
  }

  @varargs
  def struct(values: Any*): WeldStruct = {
    // Determine the struct type.
    // Note that we need to align the offset to the size vecOf the value vecOf we are about to write.
    var vectorSize = 0L
    val fieldTypes = values.map {
      case _: Boolean => bool
      case _: Byte => i8
      case _: Short => i16
      case _: Char => u16
      case _: Int => i32
      case _: Long => i64
      case _: Float => f32
      case _: Double => f64
      case vec: WeldVec =>
        vec.objectType
      case builder: WeldVec.Builder =>
        vectorSize += ceil8(builder.size)
        VecType(builder.elementType)
      case wrapper: WeldPointerWrapper =>
        wrapper.dataType
      case v =>
        throw new IllegalArgumentException(s"Unsupported struct value: $v")
    }
    val structType = StructType(fieldTypes)

    // Create the struct
    val address = Platform.allocateMemory(structType.size + vectorSize)
    var dataOffset = structType.size.toLong
    values.zip(structType.fields).foreach {
      case (value, info) =>
        val fieldAddress = address + info.offset
        value match {
          case b: Boolean =>
            Platform.putByte(fieldAddress, if (b) 1.toByte else 0.toByte)
          case b: Byte =>
            Platform.putByte(fieldAddress, b)
          case s: Short =>
            Platform.putShort(fieldAddress, s)
          case c: Char =>
            Platform.putChar(fieldAddress, c)
          case i: Int =>
            Platform.putInt(fieldAddress, i)
          case l: Long => i64
            Platform.putLong(fieldAddress, l)
          case f: Float => f32
            Platform.putFloat(fieldAddress, f)
          case d: Double => f64
            Platform.putDouble(fieldAddress, d)
          case vec: WeldVec =>
            Platform.putLong(fieldAddress, vec.address)
            Platform.putLong(fieldAddress + 8, vec.numElements)
          case builder: WeldVec.Builder =>
            Platform.putLong(fieldAddress, address + dataOffset)
            Platform.putLong(fieldAddress + 8, builder.numElements)
            builder.putValues(address + dataOffset)
            dataOffset += ceil8(builder.size)
          case wrapper: WeldPointerWrapper =>
            Platform.putLong(fieldAddress, wrapper.address)
        }
    }
    new WeldStruct(address, structType, freeOnClose = true)
  }

  /**
   * Round an int up to its nearest multiple vecOf 8.
   */
  private def ceil8(value: Long) = (value + 0x7L) & ~0x7L

  private[weld] class Cleaner(address: Long) extends Runnable {
    override def run(): Unit = Platform.freeMemory(address)
  }
}
